rfcs/0042-bitswap.md

```diff
@@ -0,0 +1,268 @@
+    - we could compute this as:
+      - **NB:** _Please double check my math above. History has proven I am not the best mathematician, and I could have gotten something wrong with my above math (would write some tests to confirm it when we go to implement it)._
+      - `let base_block_size = max_block_size - max_overhead_size`
+      - `let leftover_space s = s mod base_block_size`
+      - `let num_blocks s = s / base_block_size + if leftover_space s = 0 then 0 else 1`
+      - `let blob_cids_size = num_blocks blob_size - 1 * cid_size` (we subtract 1 here because we do not store a reference to the root block in a block tree)
+      - `let num_overflow_cids_size = min 0 (blob_cid_size - leftover_space blob_size)`
+      - `num_blocks blob_size + num_blocks num_overflow_cids_size`
+- generate a block tree layout based on how many CIDs we need to store to refer to all the blocks (minus the root block)
+- fill in slices of binary data, starting at the leaves; continue upward, propagating CIDs of child blocks to parents for inclusion in their representation)
+- at the end, you have a set of blocks, with a single block being marked as the "root" block for the set
+
+### Bitswap Block Cache
+
+There exist some key constraints in choosing a good solution for the Bitswap block cache. Importantly, the cache needs to support a concurrent reader and writer at the same time. Additionally, the cache should be optimized for existence checks and reads, as these are the operations we will perform the most frequently against the cache. The cache should have a very short delay between when data is written and when it is visible to a reader, so that the libp2p helper process has that information available to serve to other nodes on the network upon broadcast of the root block CIDs. Finally, the cache would ideally be persisted, so that we can quickly reload the cache in the event of node crashes (we want to avoid increasing bootstrap time for a node, as that keeps stake offline after crashes).
+
+Given these constraints, [LMDB](http://www.lmdb.tech/doc/) is a good choice for the Bitswap block cache. It meets all of the above criteria, including persistence.
+
+#### Cache Schema
+
+The Bitswap block cache stores 2 key pieces of information: a set of blocks that are available to serve over the network, and a mapping from Mina CIDs and Bitswap block CIDs. To be specific, most resources we want to serve over Bitswap already have content identifiers (hashes) that are computed in the daemon and referenced in the blockchain state. The Bitswap block cache will provide a mapping between these existing CIDs and the set of Bitswap block CIDs that is projected from the resource, so that the Bitswap blocks do not need to be all recomputed in the event of restarts, and we do not need to compute these Bitswap blocks in order to determine what resources are stored in the cache.
```

Comment:
what exactly are the mina CIDs? is this analogous to block hash, etc?

Reply:
Yes. I just used the term "Mina CID" to be a little more generic, as we also have, for example, existing transaction hashes to identify transactions.